scripts/release_train.py

```diff
@@ -0,0 +1,224 @@
+                               beta       47         47          agent-image (r30)
+                               edge       47         47          agent-image (r31)
+1.0      ubuntu 20.04 (amd64)  stable     50         50          agent-image (r30)
+                               candidate  56         56          agent-image (r31) <<!>>
+                               beta       56         56          agent-image (r31)
+                               edge       56         56          agent-image (r31)
+```
+
+Suppose you now release a new `edge` version on 1.0 and then run `release_train.py foo --stable`;
+now you should see (note the changes ` <<!>>`):
+```
+$ charmcraft status foo
+Track    Base                  Channel    Version    Revision    Resources
+latest   ubuntu 20.04 (amd64)  stable     47         47          agent-image (r30) <<!>>
+                               candidate  47         47          agent-image (r30)
+                               beta       47         47          agent-image (r30)
+                               edge       47         47          agent-image (r31)
+1.0      ubuntu 20.04 (amd64)  stable     56         56          agent-image (r31) <<!>>
+                               candidate  56         56          agent-image (r31)
+                               beta       57         57          agent-image (r32) <<!>>
+                               edge       57         57          agent-image (r32) <<!>>
+```
+"""
+
+import json
+import logging
+import shlex
+from collections import defaultdict
+from subprocess import Popen, PIPE
+from typing import Optional, List
+
+import typer
+
+logger = logging.getLogger(__name__)
+
+bump_train = {
+    'edge': 'beta',
+    'beta': 'candidate',
+    'candidate': 'stable',
+    'stable': None
+}
+
+
+def _run_cmd(cmd: str):
+    print(f"executing {cmd}")
+    proc = Popen(shlex.split(cmd), stdout=PIPE, text=True)
+    proc.wait()
+
+
+def _get_json(cmd: str):
+    cmd += ' --format=json'
+    proc = Popen(shlex.split(cmd), stdout=PIPE, text=True)
+    proc.wait()
+    out = proc.stdout.read()
+    return json.loads(out)
+
+
+def get_release_info(charm: str):
+    logger.debug(f"gathering charmcraft status for {charm}")
+    cmd = f'charmcraft status {charm}'
+    status = _get_json(cmd)
+    out = {}
+    for track in status:
+        releases_per_base = {}
+        track_name = track['track']
+        out[track_name] = releases_per_base
+
+        logger.debug(f"processing track {track_name}")
+
+        for mapping in track['mappings']:
+            base = mapping['base']
+            base_name = f"{base['name']}-{base['channel']}-{base['architecture']}"
+            releases_to_bump = {}
+            releases_per_base[base_name] = releases_to_bump
+
+            logger.debug(f"processing base {base}")
+            releases = mapping['releases']
+            for release in releases:
+                for elem, next_ in bump_train.items():
```

Comment:
- `for k, v` is idiomatic.
- This is the only usage so could inline it here, e.g. `for from, to in (("edge", "beta"), ...)`